### ISSUE TITLE
[RyuJIT] Morph produces a TYP_REF/TYP_BOOL compare

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#14280]
When compiling `System.Collections.Generic.ObjectEqualityComparer``1[Nullable``1][System.Nullable``1[System.__Canon]]:LastIndexOf(ref,struct,int,int):int:this` morph produces an odd looking compare while removing a box operation:
```
fgMorphTree BB09, stmt 15 (before)
               [000098] --CXG-------             *  JTRUE     void  
               [000096] ------------             |  /--*  CNS_INT   ref    null
               [000097] --CXG-------             \--*  EQ        int   
               [000095] --CXG-------                \--*  CALL help ref    HELPER.CORINFO_HELP_BOX_NULLABLE
               [000092] ------------ arg0              +--*  LCL_VAR   long   V09 tmp1         
...
GenTreeNode creates assertion:
               [000289] ---X--------             *  ARR_LENGTH int   
In BB09 New Local Constant Assertion: V01 != null index=#02, mask=0000000000000002

fgMorphTree BB09, stmt 15 (after)
               [000098] ---XG+------             *  JTRUE     void  
               [000096] -----+------             |  /--*  CNS_INT   ref    null
               [000097] J--XG+-N----             \--*  EQ        int   
               [000299] R---G+------                |  /--*  IND       bool  
               [000298] ----G-------                |  |  \--*  ADDR      byref 
...
```
This only works because lowering fixes such compares to be TYP_INT/TYP_INT or TYP_LONG/TYP_LONG compares. But the JIT should not generate such compares in the first place.

**Comment by mikedn** [comment]
Fix available in #14281

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib

**Comment by mikedn** [comment]
Fixed in #14281
